include/PetriEngine/Reachability/ReachabilitySearch.h

```diff
@@ -196,6 +206,100 @@ namespace PetriEngine {
             return false;
         }
+
+        template<typename W, typename G>
+        bool ReachabilitySearch::tryReachRandomWalk(std::vector<std::shared_ptr<PQL::Condition> >& queries,
+                                                    std::vector<ResultPrinter::Result>& results, bool usequeries,
+                                                    bool printstats, size_t maxSteps, size_t seed)
+        {
+            // set up state
+            searchstate_t ss;
+            ss.enabledTransitionsCount.resize(_net.numberOfTransitions(), 0);
+            ss.expandedStates = 0;
+            ss.exploredStates = 1;
+            ss.heurquery = queries.size() >= 2 ? std::rand() % queries.size() : 0;
+            ss.usequeries = usequeries;
+
+            // set up working area
+            Structures::State state;
+            Structures::State working;
+            _initial.setMarking(_net.makeInitialMarking());
+            state.setMarking(_net.makeInitialMarking());
+            working.setMarking(_net.makeInitialMarking());
+
+            W states(_net, _kbound);                        // stateset
+            Structures::RandomWalkPotencyQueue queue(seed); // working queue
+            G generator = _makeSucGen<G>(_net, queries);    // successor generator
+            auto r = states.add(state);
+            // this can fail due to reductions; we push tokens around and violate K
+            if(r.first){
+                // add initial to states, check queries on initial state
+                _satisfyingMarking = r.second;
+                size_t _initialMarking = r.second;
+                // check initial marking
+                if(ss.usequeries)
+                {
+                    if(checkQueries(queries, results, working, ss, &states))
+                    {
+                        if(printstats)
+                            printStats(ss, &states);
+                        _max_tokens = states.maxTokens();
+                        return true;
+                    }
+                }
+                // Typically a timeout
+                bool stopRandomWalkCondition = false;
+                while(!stopRandomWalkCondition) {
+                    // add initial to queue
+                    {
+                        PQL::DistanceContext dc(&_net, _initial.marking());
+                        queue.push(r.second, &dc, queries[ss.heurquery].get());
+                    }
+
+                    // Search! Each step is a random walk and resets the queue
+                    for(size_t stepCounter = 0, auto nid = queue.pop(); nid != Structures::Queue::EMPTY && stepCounter < maxSteps; nid = queue.pop(), ++stepCounter) {
+                        states.decode(state, nid);
+                        generator.prepare(&state);
+
+                        while(generator.next(working)) {
+                            ss.enabledTransitionsCount[generator.fired()]++;
+                            auto res = states.add(working);
```

Comment:
To generate the next working states and to check the satisfiability of the query

Reply:
Lines 266-268 should be deleted.  You should *always* push to the queue in the random walk.